### ISSUE TITLE
How do you distinguish between epochs and iterations? / How to log Val Loss with VisualDL?

### ISSUE THREAD

**Issue by khaep** [PaddlePaddle/PaddleSeg#1722]
Hey there,

after training some Semantic Segmentation networks, I am not sure how to distinguish between epochs and iterations.
For example, the number of iterations are specified in most of the [config.yml](https://github-com.translate.goog/PaddlePaddle/PaddleSeg/blob/8a9196177c9e4b0187f90cbc6a6bc38a652eba73/configs/bisenet/bisenet_cityscapes_1024x1024_160k.yml?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de#L21) files.
But the console output prints epochs and iterations (see list below).

Can you please explain the correlation between epochs and iters in PaddleSeg?

```bash
2022-01-04 19:24:33 [INFO]	[TRAIN] epoch: 1, iter: 10/100000, loss: 2.7599, lr: 0.049996, batch_cost: 0.2763, reader_cost: 0.02104, ips: 28.9558 samples/sec | ETA 07:40:25
2022-01-04 19:24:35 [INFO]	[TRAIN] epoch: 1, iter: 20/100000, loss: 2.1158, lr: 0.049991, batch_cost: 0.2270, reader_cost: 0.03153, ips: 35.2452 samples/sec | ETA 06:18:13
2022-01-04 19:24:38 [INFO]	[TRAIN] epoch: 1, iter: 30/100000, loss: 1.6946, lr: 0.049987, batch_cost: 0.2456, reader_cost: 0.04792, ips: 32.5739 samples/sec | ETA 06:49:12
2022-01-04 19:24:40 [INFO]	[TRAIN] epoch: 1, iter: 40/100000, loss: 1.5874, lr: 0.049982, batch_cost: 0.2453, reader_cost: 0.04958, ips: 32.6186 samples/sec | ETA 06:48:36
2022-01-04 19:24:43 [INFO]	[TRAIN] epoch: 1, iter: 50/100000, loss: 1.3311, lr: 0.049978, batch_cost: 0.2457, reader_cost: 0.04675, ips: 32.5536 samples/sec | ETA 06:49:22
2022-01-04 19:24:45 [INFO]	[TRAIN] epoch: 1, iter: 60/100000, loss: 1.3802, lr: 0.049973, batch_cost: 0.2465, reader_cost: 0.04984, ips: 32.4603 samples/sec | ETA 06:50:30
2022-01-04 19:24:47 [INFO]	[TRAIN] epoch: 1, iter: 70/100000, loss: 1.2869, lr: 0.049969, batch_cost: 0.2445, reader_cost: 0.04882, ips: 32.7160 samples/sec | ETA 06:47:15
2022-01-04 19:24:50 [INFO]	[TRAIN] epoch: 1, iter: 80/100000, loss: 1.3178, lr: 0.049964, batch_cost: 0.2440, reader_cost: 0.04876, ips: 32.7932 samples/sec | ETA 06:46:15
2022-01-04 19:24:52 [INFO]	[TRAIN] epoch: 1, iter: 90/100000, loss: 1.2249, lr: 0.049960, batch_cost: 0.2461, reader_cost: 0.04864, ips: 32.5126 samples/sec | ETA 06:49:43
2022-01-04 19:24:55 [INFO]	[TRAIN] epoch: 1, iter: 100/100000, loss: 1.2916, lr: 0.049955, batch_cost: 0.2532, reader_cost: 0.05190, ips: 31.5969 samples/sec | ETA 07:01:33
2022-01-04 19:24:57 [INFO]	[TRAIN] epoch: 1, iter: 110/100000, loss: 1.2852, lr: 0.049951, batch_cost: 0.2449, reader_cost: 0.04530, ips: 32.6650 samples/sec | ETA 06:47:44
2022-01-04 19:25:00 [INFO]	[TRAIN] epoch: 1, iter: 120/100000, loss: 1.1855, lr: 0.049946, batch_cost: 0.2465, reader_cost: 0.04781, ips: 32.4482 samples/sec | ETA 06:50:25
2022-01-04 19:25:02 [INFO]	[TRAIN] epoch: 1, iter: 130/100000, loss: 1.1867, lr: 0.049942, batch_cost: 0.2467, reader_cost: 0.04778, ips: 32.4259 samples/sec | ETA 06:50:39
2022-01-04 19:25:05 [INFO]	[TRAIN] epoch: 1, iter: 140/100000, loss: 1.0851, lr: 0.049937, batch_cost: 0.2514, reader_cost: 0.05101, ips: 31.8225 samples/sec | ETA 06:58:24
2022-01-04 19:25:07 [INFO]	[TRAIN] epoch: 1, iter: 150/100000, loss: 1.2347, lr: 0.049933, batch_cost: 0.2406, reader_cost: 0.04324, ips: 33.2457 samples/sec | ETA 06:40:27
2022-01-04 19:25:10 [INFO]	[TRAIN] epoch: 1, iter: 160/100000, loss: 1.1873, lr: 0.049928, batch_cost: 0.2481, reader_cost: 0.05124, ips: 32.2490 samples/sec | ETA 06:52:47
2022-01-04 19:25:12 [INFO]	[TRAIN] epoch: 1, iter: 170/100000, loss: 1.1918, lr: 0.049924, batch_cost: 0.2435, reader_cost: 0.04699, ips: 32.8513 samples/sec | ETA 06:45:10
2022-01-04 19:25:15 [INFO]	[TRAIN] epoch: 1, iter: 180/100000, loss: 1.2348, lr: 0.049919, batch_cost: 0.2461, reader_cost: 0.04819, ips: 32.5042 samples/sec | ETA 06:49:27
2022-01-04 19:25:17 [INFO]	[TRAIN] epoch: 1, iter: 190/100000, loss: 1.3164, lr: 0.049915, batch_cost: 0.2448, reader_cost: 0.04504, ips: 32.6825 samples/sec | ETA 06:47:11
2022-01-04 19:25:20 [INFO]	[TRAIN] epoch: 2, iter: 200/100000, loss: 1.0544, lr: 0.049910, batch_cost: 0.2559, reader_cost: 0.05568, ips: 31.2671 samples/sec | ETA 07:05:34
2022-01-04 19:25:22 [INFO]	[TRAIN] epoch: 2, iter: 210/100000, loss: 1.1511, lr: 0.049906, batch_cost: 0.2459, reader_cost: 0.04827, ips: 32.5279 samples/sec | ETA 06:49:02
2022-01-04 19:25:25 [INFO]	[TRAIN] epoch: 2, iter: 220/100000, loss: 1.0025, lr: 0.049901, batch_cost: 0.2462, reader_cost: 0.04900, ips: 32.4978 samples/sec | ETA 06:49:22
2022-01-04 19:25:27 [INFO]	[TRAIN] epoch: 2, iter: 230/100000, loss: 1.1605, lr: 0.049897, batch_cost: 0.2475, reader_cost: 0.04933, ips: 32.3178 samples/sec | ETA 06:51:37
2022-01-04 19:25:29 [INFO]	[TRAIN] epoch: 2, iter: 240/100000, loss: 1.0372, lr: 0.049892, batch_cost: 0.2455, reader_cost: 0.04794, ips: 32.5897 samples/sec | ETA 06:48:08
2022-01-04 19:25:32 [INFO]	[TRAIN] epoch: 2, iter: 250/100000, loss: 1.1097, lr: 0.049888, batch_cost: 0.2439, reader_cost: 0.04631, ips: 32.8041 samples/sec | ETA 06:45:26
2022-01-04 19:25:34 [INFO]	[TRAIN] epoch: 2, iter: 260/100000, loss: 1.1442, lr: 0.049883, batch_cost: 0.2479, reader_cost: 0.04898, ips: 32.2688 samples/sec | ETA 06:52:07
2022-01-04 19:25:37 [INFO]	[TRAIN] epoch: 2, iter: 270/100000, loss: 1.0965, lr: 0.049879, batch_cost: 0.2522, reader_cost: 0.05306, ips: 31.7176 samples/sec | ETA 06:59:14
2022-01-04 19:25:39 [INFO]	[TRAIN] epoch: 2, iter: 280/100000, loss: 1.0116, lr: 0.049874, batch_cost: 0.2454, reader_cost: 0.04634, ips: 32.6039 samples/sec | ETA 06:47:48
```

Furthermore a second question.
How can I print/log the validation loss in VisualDL?
It may be necessary to compare the train and val loss to identify overfitting. 
Alternatively, the comparison between train and val accuracy also can be used for this purpose.
Please give me a hint.

Kind regards

![Screenshot from 2021-11-18 07-34-58](https://user-images.githubusercontent.com/66534155/149803928-10b41fee-44ec-4570-b2c8-506cebb380ca.png)

![Screenshot from 2021-11-18 07-35-42](https://user-images.githubusercontent.com/66534155/149803736-7803526d-709b-48c1-b8dc-588ebbd9b547.png)
?

**Comment by juncaipeng** [comment]
For the first question, `iters = gpu_num * batch_size * epoch`.

For the second question, PaddleSeg does not calculate the loss of evaluation. You should set `do_eval` as follows for `train.py`, PaddleSeg will run evaluating in training stage, write the val accuracy to VisualDL, save the model with the highest accuracy for validation dataset.

![image](https://user-images.githubusercontent.com/52520497/149872056-d1d927cc-140e-415a-93cd-fbae9b2ec595.png)

**Comment by khaep** [comment]
Thank you for the second question.

The answer of the first question is a little bit confusing. 
I trained BiSeNetV2 with 
* batch_size = 8
*  iters = 100000
* one single GPU

If I use your equation to calculate the number of epochs:
epoch = iters/(gpu_num * batch_size) = 100000/(1 * 8) = 12500

But the console output says 511 epochs.
```bash
2022-01-05 02:17:57 [INFO]	[TRAIN] epoch: 511, iter: 100000/100000, loss: 0.3656, lr: 0.000002, batch_cost: 0.2380, reader_cost: 0.04746, ips: 33.6113 samples/sec | ETA 00:00:00
2022-01-05 02:17:57 [INFO]	Start evaluating (total_samples: 104, total_iters: 104)...
```

That does not match with the equation.
Can you explain that please?

**Comment by juncaipeng** [comment]
Sorry, it should be `epoch = (iters * gpu_num * batch_size) / dataset_num `.

**Comment by khaep** [comment]
Okay,
that means dataset_num is the number of training images.

So for the case mentioned above with 1570 training images:
epoch = (100000 * 1 * 8) / 1570 = 509.55 -> approximately 511

Is that correct?

**Comment by wuyefeilin** [comment]
Yes, it is correct. Because the last iter of one epoch may be ignored.
